examples/deep_learning/distributed_flax_imagenet.py

```diff
@@ -0,0 +1,839 @@
+    model_dtype = jnp.bfloat16 if platform == 'tpu' else jnp.float16
+  return model_cls(num_classes=num_classes, dtype=model_dtype, **kwargs)
+
+
+def initialize_model(
+    key: PRNGKey,
+    model: nn.Module,
+    image_size: int = IMAGE_SIZE,
+) -> Tuple[ArrayTree, ArrayTree]:
+  """Initializes FLAX model, returning params and mutable variables."""
+  input_shape = (1, image_size, image_size, 3)
+  @jax.jit
+  def init(*args):
+    return model.init(*args)
+  variables = init({'params': key}, jnp.ones(input_shape, model.dtype))
+
+  return variables['params'], variables['batch_stats']
+
+
+def create_learning_rate_fn(
+    learning_rate: float,
+    batch_size: int,
+    steps_per_epoch: int,
+    warmup_epochs: int,
+    num_epochs: int,
+) -> LearningRateFn:
+  """Creates learning rate schedule."""
+  base_learning_rate = learning_rate * batch_size / 256.
+  warmup_fn = optax.linear_schedule(
+      init_value=0., end_value=base_learning_rate,
+      transition_steps=warmup_epochs * steps_per_epoch)
+  cosine_epochs = max(num_epochs - warmup_epochs, 1)
+  cosine_fn = optax.cosine_decay_schedule(
+      init_value=base_learning_rate,
+      decay_steps=cosine_epochs * steps_per_epoch)
+  schedule_fn = optax.join_schedules(
+      schedules=[warmup_fn, cosine_fn],
+      boundaries=[warmup_epochs * steps_per_epoch])
+  return schedule_fn
+
+
+def create_solver(
+    learning_rate_fn: LearningRateFn,
+    momentum: float,
+    model: nn.Module,
+) -> jaxopt.OptaxSolver:
+  """Creates JAXopt solver."""
+  opt = optax.sgd(learning_rate=learning_rate_fn,
+                  momentum=momentum,
+                  nesterov=True)
+  fun = functools.partial(loss_fun, model=model)
+  return jaxopt.OptaxSolver(opt=opt, fun=fun, has_aux=True)
+
+
+def zeros_like_fun_output(
+    fun: Callable,
+    index: Optional[int] = None,
+) -> Callable:
+  """Replaces fun, outputting a pytree of zeroes with the original structure."""
+  def wrapper(*args, **kwargs):
+    pytree = jax.eval_shape(fun, *args, **kwargs)
+    leaves, treedef = jax.tree_flatten(pytree)
+    leaves = [jnp.zeros(shape=leaf.shape, dtype=leaf.dtype) for leaf in leaves]
+    zeros_like_pytree = jax.tree_unflatten(treedef, leaves)
+    return zeros_like_pytree if index is None else zeros_like_pytree[index]
+  return wrapper
+
+
+def initialize_solver(
+    solver: jaxopt.OptaxSolver,
+    init_params: ArrayTree,
+    init_batch_stats: ArrayTree,
+    first_batch: Batch,
+    model: nn.Module,
+) -> ArrayTree:
+  """Initializes the state of jaxopt.OptaxSolver."""
+  # "Default" JAXopt initial optimizer state.
+  state = solver.init_state(init_params)
+
+  # To prevent `train_step` from being compiled twice, we must ensure all its
+  # input arguments have the same shape and dtype in all calls. To this end,
+  # we will
+  #   1) Initialize `state.aux` with a Pytree of the right shape and dtype.
+  #   2) Ensure that `state.value` and `state.error` are strongly typed.
```

Comment:
This requires a lot of low-level knowledge from the user. In the future, it would be nice if we can find a way to avoid this. CC @froystig

Reply:
The `initialize_solver` and `initialize_metrics` functions could be somewhat less "intimidating" if I had hard-coded the initialization, e.g.,
```
init_aux = LossFnAux(batch_stats=init_batch_stats, logits=jnp.zeros([batch_size, NUM_CLASSES], dtype=jnp.bfloat16))
```
However, I wanted to use this example to see if we could infer the initialization of `state.aux` and `metrics` automatically instead of asking the user to specify it. These functions show we can, but there might certainly be better ways to accomplish the same.

Going forward, I anticipate these functions (or an improved version of them) could be part of the future `solver.run` / `solver.run_iterator` API. But I do wonder if we should make them somehow accessible to users writing their own manual training loops.